allennlp/training/util.py

```diff
@@ -428,38 +415,43 @@ def make_vocab_from_params(
             "The 'vocabulary' directory in the provided serialization directory is non-empty"
         )
 
-    all_datasets = datasets_from_params(params)
-    datasets_for_vocab_creation = set(params.pop("datasets_for_vocab_creation", all_datasets))
-
-    for dataset in datasets_for_vocab_creation:
-        if dataset not in all_datasets:
-            raise ConfigurationError(f"invalid 'dataset_for_vocab_creation' {dataset}")
-
-    logger.info(
-        "From dataset instances, %s will be considered for vocabulary creation.",
-        ", ".join(datasets_for_vocab_creation),
+    datasets_for_vocab_creation: Optional[List[str]] = params.pop(
+        "datasets_for_vocab_creation", None
     )
+    datasets: Dict[str, Dataset]
+    if datasets_for_vocab_creation is None:
+        # If `datasets_for_vocab_creation` was not specified, we'll use all datasets
+        # from the config.
+        datasets = datasets_from_params(params)
+    else:
+        for dataset in datasets_for_vocab_creation:
+            data_path = f"{dataset}_data_path"
+            if data_path not in params:
+                raise ConfigurationError(f"invalid 'dataset_for_vocab_creation' {dataset}")
+        datasets = datasets_from_params(
+            params,
+            train=("train" in datasets_for_vocab_creation),
+            validation=("validation" in datasets_for_vocab_creation),
+            test=("test" in datasets_for_vocab_creation),
+        )
 
     instances: Iterable[Instance] = (
         instance
-        for key, dataset in all_datasets.items()
-        if key in datasets_for_vocab_creation
+        for key, dataset in datasets.items()
+        if datasets_for_vocab_creation is None or key in datasets_for_vocab_creation
         for instance in dataset
     )
 
-    if print_statistics:
```

Comment:
I'm pretty sure this was here because in a prior version of the code, `instances` was a generator that could get consumed when constructing the vocabulary, and then nothing would get printed.  I assume that's no longer an issue?  I've lost track of how the datasets work now.

Reply:
Yea, idk. It seems like a hack to turn the dataset into a big batch just to print statistics.